build.gradle

```diff
@@ -140,6 +140,8 @@ subprojects {
       }
     })
 
+    maxHeapSize = "1500m"
```

Comment:
I had to set this as occasionally some of the ORC tests hit an OOM. Didn't dig into it too much, as `maxHeapSize` is already set for one specific project.

Will remove post-review if we only use parquet.

Reply:
Would you mind to share you OOM stacktrace or some detailed log informations ?  I think we need to take a look because we've just run so many times of those unit tests, and we rarely encountered an ORC OOM. I'd like to check whether this change lead to this OOM issue.

Reply:
Sure. I’ll need to see if I can reproduce it.

I think that adding class unloading and a specific maximum for max meta space size would help, but configuring that was difficult.

The stack trace only came up in the new upsert tests, which uses many catalogs and file formats. When running the whole test suite locally, it would happen on ORC.

I’ll get you a stack trace shortly. 👍

Reply:
I am not able to reproduce it now, but I originally encountered this running locally using
`./gradlew clean build`

I had some of the stack trace in my terminal history still though:
```
                        Caused by:
                        java.lang.OutOfMemoryError: GC overhead limit exceeded
                            at java.util.Hashtable.clone(Hashtable.java:548)
                            at org.apache.hadoop.conf.Configuration.<init>(Configuration.java:706)
                            at org.apache.iceberg.orc.ORC$WriteBuilder.<init>(ORC.java:91)
                            at org.apache.iceberg.orc.ORC$WriteBuilder.<init>(ORC.java:79)
                            at org.apache.iceberg.orc.ORC.write(ORC.java:76)
                            at org.apache.iceberg.orc.ORC$DeleteWriteBuilder.<init>(ORC.java:333)
                            at org.apache.iceberg.orc.ORC$DeleteWriteBuilder.<init>(ORC.java:320)
                            at org.apache.iceberg.orc.ORC.writeDeletes(ORC.java:317)
                            at org.apache.iceberg.flink.sink.FlinkAppenderFactory.newEqDeleteWriter(FlinkAppenderFactory.java:166)
                            at org.apache.iceberg.io.BaseTaskWriter$RollingEqDeleteWriter.newWriter(BaseTaskWriter.java:348)
                            at org.apache.iceberg.io.BaseTaskWriter$RollingEqDeleteWriter.newWriter(BaseTaskWriter.java:341)
                            at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.openCurrent(BaseTaskWriter.java:279)
                            at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.<init>(BaseTaskWriter.java:241)
                            at org.apache.iceberg.io.BaseTaskWriter$BaseRollingWriter.<init>(BaseTaskWriter.java:231)
                            at org.apache.iceberg.io.BaseTaskWriter$RollingEqDeleteWriter.<init>(BaseTaskWriter.java:343)
                            at org.apache.iceberg.io.BaseTaskWriter$BaseEqualityDeltaWriter.<init>(BaseTaskWriter.java:108)
                            at org.apache.iceberg.flink.sink.BaseDeltaTaskWriter$RowDataDeltaWriter.<init>(BaseDeltaTaskWriter.java:105)
                            at org.apache.iceberg.flink.sink.PartitionedDeltaWriter.route(PartitionedDeltaWriter.java:67)
                            at org.apache.iceberg.flink.sink.BaseDeltaTaskWriter.write(BaseDeltaTaskWriter.java:77)
                            at org.apache.iceberg.flink.sink.BaseDeltaTaskWriter.write(BaseDeltaTaskWriter.java:41)
                            at org.apache.iceberg.flink.sink.IcebergStreamWriter.processElement(IcebergStreamWriter.java:72)
                            at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask$StreamTaskNetworkOutput.emitRecord(OneInputStreamTask.java:233)
                            at org.apache.flink.streaming.api.operators.sort.SortingDataInput.emitNextSortedRecord(SortingDataInput.java:210)
                            at org.apache.flink.streaming.api.operators.sort.SortingDataInput.emitNext(SortingDataInput.java:190)
                            at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
                            at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:496)
                            at org.apache.flink.streaming.runtime.tasks.StreamTask$$Lambda$2138/1830483331.runDefaultAction(Unknown Source)
                            at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:203)
                            at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:809)
                            at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:761)
                            at org.apache.flink.runtime.taskmanager.Task$$Lambda$2190/1644560384.run(Unknown Source)
                            at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:958)
```

Reply:
Spoke too soon. By commenting this line out, and then running `./gradlew clean iceberg-flink:iceberg-flink-1.14:test` I got the full error.

```
> Task :iceberg-flink:iceberg-flink-1.14:test

org.apache.iceberg.flink.TestFlinkUpsert > testPrimaryKeyKEqualToPartitionKey[catalogName=testhive, baseNamespace=, format=ORC, isStreaming=false] FAILED
    java.lang.RuntimeException: Failed to collect table result
        at org.apache.iceberg.flink.FlinkTestBase.sql(FlinkTestBase.java:102)
        at org.apache.iceberg.flink.TestFlinkUpsert.testPrimaryKeyKEqualToPartitionKey(TestFlinkUpsert.java:182)

        Caused by:
        java.lang.RuntimeException: Failed to fetch next result
            at org.apache.flink.streaming.api.operators.collect.CollectResultIterator.nextResultFromFetcher(CollectResultIterator.java:109)
            at org.apache.flink.streaming.api.operators.collect.CollectResultIterator.hasNext(CollectResultIterator.java:80)
            at org.apache.flink.table.api.internal.TableResultImpl$CloseableRowIteratorWrapper.hasNext(TableResultImpl.java:370)
            at org.apache.iceberg.relocated.com.google.common.collect.Iterators.addAll(Iterators.java:366)
            at org.apache.iceberg.relocated.com.google.common.collect.Lists.newArrayList(Lists.java:147)
            at org.apache.iceberg.flink.FlinkTestBase.sql(FlinkTestBase.java:100)
            ... 1 more

            Caused by:
            java.io.IOException: Failed to fetch job execution result
                at org.apache.flink.streaming.api.operators.collect.CollectResultFetcher.getAccumulatorResults(CollectResultFetcher.java:177)
                at org.apache.flink.streaming.api.operators.collect.CollectResultFetcher.next(CollectResultFetcher.java:120)
                at org.apache.flink.streaming.api.operators.collect.CollectResultIterator.nextResultFromFetcher(CollectResultIterator.java:106)
                ... 6 more

                Caused by:
                java.util.concurrent.ExecutionException: org.apache.flink.runtime.client.JobExecutionException: Job execution failed.
                    at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
                    at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928)
                    at org.apache.flink.streaming.api.operators.collect.CollectResultFetcher.getAccumulatorResults(CollectResultFetcher.java:175)
                    ... 8 more

                    Caused by:
                    org.apache.flink.runtime.client.JobExecutionException: Job execution failed.
                        at org.apache.flink.runtime.jobmaster.JobResult.toJobExecutionResult(JobResult.java:144)
                        at org.apache.flink.runtime.minicluster.MiniClusterJobClient.lambda$getJobExecutionResult$3(MiniClusterJobClient.java:137)
                        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
                        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591)
                        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
                        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
                        at org.apache.flink.runtime.rpc.akka.AkkaInvocationHandler.lambda$invokeRpc$0(AkkaInvocationHandler.java:250)
                        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
                        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
                        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
                        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
                        at org.apache.flink.util.concurrent.FutureUtils.doForward(FutureUtils.java:1389)
                        at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$null$1(ClassLoadingUtils.java:93)
                        at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68)
                        at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$guardCompletionWithContextClassLoader$2(ClassLoadingUtils.java:92)
                        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
                        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
                        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
                        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
                        at org.apache.flink.runtime.concurrent.akka.AkkaFutureUtils$1.onComplete(AkkaFutureUtils.java:47)
                        at akka.dispatch.OnComplete.internal(Future.scala:300)
                        at akka.dispatch.OnComplete.internal(Future.scala:297)
                        at akka.dispatch.japi$CallbackBridge.apply(Future.scala:224)
                        at akka.dispatch.japi$CallbackBridge.apply(Future.scala:221)
                        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
                        at org.apache.flink.runtime.concurrent.akka.AkkaFutureUtils$DirectExecutionContext.execute(AkkaFutureUtils.java:65)
                        at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
                        at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1(Promise.scala:284)
                        at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1$adapted(Promise.scala:284)
                        at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
                        at akka.pattern.PromiseActorRef.$bang(AskSupport.scala:621)
                        at akka.pattern.PipeToSupport$PipeableFuture$$anonfun$pipeTo$1.applyOrElse(PipeToSupport.scala:24)
                        at akka.pattern.PipeToSupport$PipeableFuture$$anonfun$pipeTo$1.applyOrElse(PipeToSupport.scala:23)
                        at scala.concurrent.Future.$anonfun$andThen$1(Future.scala:532)
                        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
                        at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
                        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
                        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
                        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
                        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
                        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
                        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
                        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
                        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
                        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
                        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
                        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
                        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)

                        Caused by:
                        org.apache.flink.runtime.JobException: Recovery is suppressed by NoRestartBackoffTimeStrategy
                            at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.handleFailure(ExecutionFailureHandler.java:138)
                            at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.getFailureHandlingResult(ExecutionFailureHandler.java:82)
                            at org.apache.flink.runtime.scheduler.DefaultScheduler.handleTaskFailure(DefaultScheduler.java:228)
                            at org.apache.flink.runtime.scheduler.DefaultScheduler.maybeHandleTaskFailure(DefaultScheduler.java:218)
                            at org.apache.flink.runtime.scheduler.DefaultScheduler.updateTaskExecutionStateInternal(DefaultScheduler.java:209)
                            at org.apache.flink.runtime.scheduler.SchedulerBase.updateTaskExecutionState(SchedulerBase.java:679)
                            at org.apache.flink.runtime.scheduler.SchedulerNG.updateTaskExecutionState(SchedulerNG.java:79)
                            at org.apache.flink.runtime.jobmaster.JobMaster.updateTaskExecutionState(JobMaster.java:444)
                            at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                            at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                            at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                            at java.lang.reflect.Method.invoke(Method.java:498)
                            at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.lambda$handleRpcInvocation$1(AkkaRpcActor.java:316)
                            at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:83)
                            at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcInvocation(AkkaRpcActor.java:314)
                            at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcMessage(AkkaRpcActor.java:217)
                            at org.apache.flink.runtime.rpc.akka.FencedAkkaRpcActor.handleRpcMessage(FencedAkkaRpcActor.java:78)
                            at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleMessage(AkkaRpcActor.java:163)
                            at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:24)
                            at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:20)
                            at scala.PartialFunction.applyOrElse(PartialFunction.scala:123)
                            at scala.PartialFunction.applyOrElse$(PartialFunction.scala:122)
                            at akka.japi.pf.UnitCaseStatement.applyOrElse(CaseStatements.scala:20)
                            at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
                            at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
                            at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
                            at akka.actor.Actor.aroundReceive(Actor.scala:537)
                            at akka.actor.Actor.aroundReceive$(Actor.scala:535)
                            at akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:220)
                            at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
                            at akka.actor.ActorCell.invoke(ActorCell.scala:548)
                            at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
                            at akka.dispatch.Mailbox.run(Mailbox.scala:231)
                            at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
                            ... 4 more

                            Caused by:
                            org.apache.iceberg.exceptions.RuntimeIOException: Failed to open file: file:/var/folders/ry/c84wjq4j2tlc6vv6yk1prqsh0000gn/T/junit2796242223497120471/db.db/upsert_on_data_key/data/data=aaa/00000-0-6d373980-4ea6-4df7-8fe6-4b8ed7c0b4f7-00001.orc
                                at org.apache.iceberg.orc.ORC.newFileReader(ORC.java:574)
                                at org.apache.iceberg.orc.ORC.newFileReader(ORC.java:583)
                                at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:78)
                                at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:45)
                                at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
                                at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:73)
                                at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:65)
                                at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
                                at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:73)
                                at org.apache.iceberg.io.CloseableIterable$4$1.<init>(CloseableIterable.java:99)
                                at org.apache.iceberg.io.CloseableIterable$4.iterator(CloseableIterable.java:98)
                                at org.apache.iceberg.flink.source.RowDataFileScanTaskReader.open(RowDataFileScanTaskReader.java:86)
                                at org.apache.iceberg.flink.source.DataIterator.openTaskIterator(DataIterator.java:136)
                                at org.apache.iceberg.flink.source.DataIterator.updateCurrentIterator(DataIterator.java:126)
                                at org.apache.iceberg.flink.source.DataIterator.hasNext(DataIterator.java:103)
                                at org.apache.iceberg.flink.source.FlinkInputFormat.reachedEnd(FlinkInputFormat.java:111)
                                at org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction.run(InputFormatSourceFunction.java:89)
                                at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:116)
                                at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:73)
                                at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:323)

                                Caused by:
                                java.io.IOException: Problem reading file footer file:/var/folders/ry/c84wjq4j2tlc6vv6yk1prqsh0000gn/T/junit2796242223497120471/db.db/upsert_on_data_key/data/data=aaa/00000-0-6d373980-4ea6-4df7-8fe6-4b8ed7c0b4f7-00001.orc
                                    at org.apache.orc.impl.ReaderImpl.extractFileTail(ReaderImpl.java:841)
                                    at org.apache.orc.impl.ReaderImpl.<init>(ReaderImpl.java:566)
                                    at org.apache.orc.OrcFile.createReader(OrcFile.java:385)
                                    at org.apache.iceberg.orc.ORC.newFileReader(ORC.java:572)
                                    ... 19 more

                                    Caused by:
                                    java.lang.OutOfMemoryError: GC overhead limit exceeded
                                        at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57)
                                        at java.nio.ByteBuffer.allocate(ByteBuffer.java:335)
                                        at org.apache.orc.impl.InStream.allocateBuffer(InStream.java:240)
                                        at org.apache.orc.impl.InStream.access$200(InStream.java:38)
                                        at org.apache.orc.impl.InStream$CompressedStream.allocateForUncompressed(InStream.java:443)
                                        at org.apache.orc.impl.InStream$CompressedStream.readHeader(InStream.java:496)
                                        at org.apache.orc.impl.InStream$CompressedStream.ensureUncompressed(InStream.java:528)
                                        at org.apache.orc.impl.InStream$CompressedStream.read(InStream.java:515)
                                        at java.io.InputStream.read(InputStream.java:101)
                                        at com.google.protobuf25.CodedInputStream.refillBuffer(CodedInputStream.java:737)
                                        at com.google.protobuf25.CodedInputStream.isAtEnd(CodedInputStream.java:701)
                                        at com.google.protobuf25.CodedInputStream.readTag(CodedInputStream.java:99)
                                        at org.apache.orc.OrcProto$Footer.<init>(OrcProto.java:27472)
                                        at org.apache.orc.OrcProto$Footer.<init>(OrcProto.java:27436)
                                        at org.apache.orc.OrcProto$Footer$1.parsePartialFrom(OrcProto.java:27612)
                                        at org.apache.orc.OrcProto$Footer$1.parsePartialFrom(OrcProto.java:27607)
                                        at com.google.protobuf25.AbstractParser.parseFrom(AbstractParser.java:89)
                                        at com.google.protobuf25.AbstractParser.parseFrom(AbstractParser.java:95)
                                        at com.google.protobuf25.AbstractParser.parseFrom(AbstractParser.java:49)
                                        at org.apache.orc.OrcProto$Footer.parseFrom(OrcProto.java:28150)
                                        at org.apache.orc.impl.ReaderImpl.extractFileTail(ReaderImpl.java:829)
                                        at org.apache.orc.impl.ReaderImpl.<init>(ReaderImpl.java:566)
                                        at org.apache.orc.OrcFile.createReader(OrcFile.java:385)
                                        at org.apache.iceberg.orc.ORC.newFileReader(ORC.java:572)
                                        at org.apache.iceberg.orc.ORC.newFileReader(ORC.java:583)
                                        at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:78)
                                        at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:45)
                                        at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
                                        at org.apache.iceberg.util.Filter$$Lambda$2657/732398219.iterator(Unknown Source)
                                        at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:73)
                                        at org.apache.iceberg.io.CloseableIterable$2.iterator(CloseableIterable.java:65)
                                        at org.apache.iceberg.util.Filter.lambda$filter$0(Filter.java:35)
```

Reply:
It should be noted that `TestSplitScan` in `iceberg-data` requires this heap size as well.

That's where I pullled the 1500m value from.

Reply:
I was also encountered a failure when I commented this line.  There must be something wrong...

```
> Task :iceberg-flink:iceberg-flink-1.14:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':iceberg-flink:iceberg-flink-1.14:test'.
> Process 'Gradle Test Executor 10' finished with non-zero exit value 239
  This problem might be caused by incorrect test process configuration.
  Please refer to the test execution section in the User Manual at https://docs.gradle.org/7.4.1/userguide/java_testing.html#sec:test_execution
```

Reply:
This seems like a separate issue. Since this is an important problem to fix, my preference is to find the memory issue later and get this PR merged.

Reply:
Agreed.